torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -5423,6 +5425,45 @@ std::shared_ptr<c10::Allocator> ProcessGroupNCCL::getMemAllocator() {
   return ncclMemAllocator;
 }
 
+at::Tensor ProcessGroupNCCL::allocateTensor(
+    long size,
+    at::TensorOptions options) {
+  // Some checks
+  TORCH_CHECK(options.has_device(), "Tensor options must include device");
+  auto device = options.device();
+  TORCH_CHECK(
+      device.is_cuda(),
+      "NCCL tensor allocator expects cuda type but got " + c10::str(device))
+
+  at::cuda::OptionalCUDAGuard gpuGuard(device);
+
+  // Create memory pool
+  if (!memPool_) {
+    // Needs a CUDAAllocator
+    auto allocator =
+        reinterpret_cast<c10::cuda::CUDACachingAllocator::CUDAAllocator*>(
+            getMemAllocator().get());
+    // Pool is created
+    memPool_ = new c10::cuda::MemPool(allocator);
```

Comment:
Should we explicitly use `new` and delete or would handling it with a unique_ptr be better?

Reply:
Yeah, that's a good idea. I used `unique_ptr` originally but realized that almost all APIs around `MemPool` takes raw pointers -- so I had to use a lot of `.get()`. Any recommendation for this situation if we use `unique_ptr`?

Reply:
@kwen2501 Annoying for the get() calls, but still better than manual memory management for a whole host of reasns.

Reply:
Thanks!

Reply:
`unique_ptr` looks good to me because ProcessGroupNCCL has ownership of the pool.